meetings/working-groups/collection-literals/CL-2023-06-12.md

```diff
@@ -0,0 +1,161 @@
+public class HashSet<T>
+{
+    // existing constructor
+    public HashSet(IEnumerable<T> values);
+
+    // newly added
+    public HashSet(ReadOnlySpan<T> values);
+}
+
+// code:
+new HashSet<int>(new int[] { 1, 2, 3 });
+```
+
+With the rules of the language today, the above is ambiguous.  This is unfortunate as a prime reason to create Span overloads of methods is to allow for efficient options for callers that do not need to use the heap.  We believe this should be fixed, and the WG will present a proposal to the LDM to add overload resolution rules that consider Spans better than items in the inheritance chain for arrays.
+
+### Shape 2: Writable `Span<T>` or `T[]` storage.
+
+While the above pattern is good for collection types that own their own storage arrangement, it still involves producing the sequence one end, and then processing it to produce the file result within the factory method.  This is unnecessary overhead for a couple of very important, widely used, collections in the .Net ecosystem.  Specifically, `List<T>` and `ImmutableArray<T>`.  Both of these types are thin wrappers around a backing array, and it is highly desirable to just be able to write directly into that array without any overhead at all.  To that end, the above attribute can be used to point to a factory method with the following shape:
+
+```c#
+[CollectionBuilder(typeof(CollectionsMarshall), "Create")]
+public struct ImmutableArray<T> { /*...*/ }
+
+// In System.Runtime.CompilerServices
+public static class CollectionsMarshall
+{
+    public static Span<T> Create<T>(int capacity, out ImmutableArray<T> result); // or:
+    public static void Create<T>(int capacity, out ImmutableArray<T> result, out Span<T> storage);
+}
+
+// code:
+ImmutableArray<string> values = ["a", "b", "c"];
+
+// translation:
+Span<string> __storage = CollectionsMarshal.Create<string>(capacity: 3, out ImmutableArray<string> values);
+__storage[0] = "a";
+__storage[1] = "b";
+__storage[2] = "c";
+```
+
+Depending on how the compiler encodes constant information, this could also be a case where the data is stored in the read-only section of the dll and memcpy'ed to the destination.
+
+Note: the runtime already has methods that expose this data for high performance scenarios.  So this does not introduce any new safety concerns.  The methods are intentionally kept out of the way in the `CompilerServices` namespace to help indicate they are not for normal use, and should be treated very carefully.
+
+#### Shape 2 consequences
+
+The above pattern will not work as efficiently in the case where the Span cannot be kept on the stack while evaluating the elements of the collection literal (for example, if there are `await` calls in the literal.  An alternative/complimentary approach that could be offered would be to have the following:
+
+```c#
+public static class CollectionsMarshall
+{
+    public static void Create<T>(int capacity, out ImmutableArray<T> result, out Span<T> storage); // or:
```

Comment:
The first overload here is also in the example above. Is that intentional?

Reply:
yes.  was showing them together so the diff could be seen.